layer/unittest/trace_event_log_tests.cc

```diff
@@ -113,9 +276,180 @@ TEST(TraceEvent, CompleteEventCreation) {
+TEST(TraceEvent, InstantEventToString) {
+  InstantEvent instant_event("compile_time_init", "compile_time", 123, 321);
+  std::ostringstream expected_str;
+  expected_str << "{ " << std::quoted("name") << " : "
+               << std::quoted("compile_time_init") << ", " << std::quoted("ph")
+               << " : " << std::quoted("i") << ", " << std::quoted("cat")
+               << " : " << std::quoted("compile_time") << ", "
+               << std::quoted("pid") << " : " << 123 << ", "
+               << std::quoted("tid") << " : " << 321 << ", "
+               << std::quoted("ts") << " : " << std::fixed
+               << ToUnixMillis(instant_event.GetCreationTime().GetValue())
+               << ", " << std::quoted("s") << " : " << std::quoted("g") << ", "
+               << std::quoted("args") << " : "
+               << "{ " << std::quoted("scope") << " : " << std::quoted("g")
+               << " } },";
+  EXPECT_EQ(EventToTraceEventString(instant_event), expected_str.str());
+}
+
+TEST(TraceEvent, CompleteEventToString) {
+  const int64_t hash_val1 = 0x67d6fd0aaa78a6d8;
+  const int64_t hash_val2 = 0x67d390249c2f20ce;
+  Duration duration = Duration::FromNanoseconds(100000);
+  CreateShaderCompleteEvent complete_event(
+      "compile_time", {hash_val1, hash_val2}, duration, 123, 321);
+
+  std::ostringstream expected_str;
+  double start_clock =
+      ToUnixMillis(complete_event.GetCreationTime().GetValue()) -
+      duration.ToMilliseconds();
+  expected_str << "{ " << std::quoted("name") << " : "
+               << std::quoted("compile_time") << ", " << std::quoted("ph")
+               << " : " << std::quoted("X") << ", " << std::quoted("cat")
+               << " : " << std::quoted("pipeline") << ", " << std::quoted("pid")
+               << " : " << 123 << ", " << std::quoted("tid") << " : " << 321
+               << ", " << std::quoted("ts") << " : " << std::fixed
+               << start_clock << ", " << std::quoted("dur") << " : "
+               << duration.ToMilliseconds() << ", " << std::quoted("args")
+               << " : "
+               << "{ " << std::quoted("duration") << " : " << std::fixed
+               << duration.ToMilliseconds() << " } },";
+  EXPECT_EQ(EventToTraceEventString(complete_event), expected_str.str());
+}
+
+// An invalid event used for checking assertions in the helper functions.
+class UnhandledTypeEvent : public Event {
+ public:
+  UnhandledTypeEvent()
+      : Event("name"),
+        trace_event_("trace_event", "category", "z", 123, 321, {}) {
+    InitAttributes({&trace_event_});
+  }
+
+ private:
+  TraceEventAttr trace_event_;
+};
+
+#ifndef NDEBUG
+TEST(TraceEvent, InvalidEventType) {
+  Event event("event without TraceEventAttr");
+  EXPECT_DEATH(EventToTraceEventString(event),
+               "Could not find TraceEventAttr in the event.");
+
+  UnhandledTypeEvent unrecognized_event = {};
+  ASSERT_DEATH(EventToTraceEventString(unrecognized_event),
+               "Unrecognized phase.");
```

Comment:
Nice! But shouldn't these be two separate test cases? Does the second one ever get executed when the first one dies?

Reply:
I think it's fine. When I change the matching string of the `ASSERT_DEATH`, it gives me errors.

Reply:
👍